README.md
# Cypress Image Snapshot

Cypress Image Snapshot binds [jest-image-snapshot](https://github.com/americanexpress/jest-image-snapshot)'s image diffing logic to [Cypress.io](https://cypress.io) commands.

[![build-and-test](https://github.com/emerson-eps/cypress-image-snapshot/actions/workflows/build-and-test.yml/badge.svg)](https://github.com/emerson-eps/cypress-image-snapshot/actions/workflows/build-and-test.yml)

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Installation](#installation)
  - [TypeScript](#typescript)
- [Usage](#usage)
  - [In your tests](#in-your-tests)
    - [Options](#options)
  - [Updating snapshots](#updating-snapshots)
  - [Preventing failures](#preventing-failures)
  - [Requiring snapshots to be present](#requiring-snapshots-to-be-present)
- [How it works](#how-it-works)
- [Requirements](#requirements)
- [Contributing](#contributing)
  - [Setup](#setup)
  - [Working on the plugin](#working-on-the-plugin)
    - [open](#open)
    - [run](#run)
      - [Note on environment variables](#note-on-environment-variables)
- [Forked from `jaredpalmer/cypress-image-snapshot`](#forked-from-jaredpalmercypress-image-snapshot)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Installation

Install with your chosen package manager

```bash
# yarn
yarn add --dev @emerson-eps/cypress-image-snapshot

# npm
npm install --save-dev @emerson-eps/cypress-image-snapshot
```

Next, import the plugin function and add it to the [`setupNodeEvents` function](https://docs.cypress.io/guides/references/configuration#setupNodeEvents):

```ts
// cypress.config.ts

import {defineConfig} from 'cypress'
import {addMatchImageSnapshotPlugin} from '@emerson-eps/cypress-image-snapshot/plugin'

export default defineConfig({
  e2e: {
    setupNodeEvents(on) {
      addMatchImageSnapshotPlugin(on)
    },
  },
})
```

Add the command to your relevant [support file](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Support-file):

```ts
// cypress/support/e2e.ts

import {addMatchImageSnapshotCommand} from '@emerson-eps/cypress-image-snapshot'

addMatchImageSnapshotCommand()

// can also add any default options to be used
// by all instances of `matchImageSnapshot`
addMatchImageSnapshotCommand({
  failureThreshold: 0.2,
})
```

### TypeScript

TypeScript is supported so any reference to `@types/cypress-image-snapshot` can be removed from your project

Ensure that the types are included in your `tsconfig.json`

```
{
  "compilerOptions": {
    // ...
  },
  "include": ["@emerson-eps/cypress-image-snapshot"]
}
```

## Usage

### In your tests

```ts
describe('Login', () => {
  it('should be publicly accessible', () => {
    cy.visit('/login');

    // snapshot name will be the test title
    cy.matchImageSnapshot();

    // snapshot name will be the name passed in
    cy.matchImageSnapshot('login');

    // options object passed in
    cy.matchImageSnapshot({
      failureThreshold: 0.4
      blur: 10
      timeout: 3000
    });

    // match element snapshot
    cy.get('#login').matchImageSnapshot();
  });
});
```

#### Options

The options object combines jest-image-snapshot and Cypress screenshot configuration.

- [jest-image-snapshot](https://github.com/americanexpress/jest-image-snapshot#%EF%B8%8F-api)
- [Cypress screenshot](https://docs.cypress.io/api/commands/screenshot#Arguments)

```ts
cy.matchImageSnapshot({
  // options for jest-image-snapshot
  failureThreshold: 0.2,
  comparisonMethod: 'ssim',

  // options for Cypress.screenshot()
  capture: 'viewport',
  blackout: ['.some-element'],
})
```

New options have been implemented to make the function recursive and to retry the snapshots:

```ts
cy.matchImageSnapshot({
  // Time it takes for the snapshot command to time out if the snapshot is not correct
  timeout: 5000,

  //Sets a delay between recursive snapshots
  delayBetweenTries: 1000,
})
```

### Updating snapshots

Run Cypress with `--env updateSnapshots=true` in order to update the base image files for all of your tests.

### Preventing failures

Run Cypress with `--env failOnSnapshotDiff=false` in order to prevent test failures when an image diff does not pass.

### Requiring snapshots to be present

Run Cypress with `--env requireSnapshots=true` in order to fail if snapshots are missing. This is useful in continuous integration where snapshots should be present in advance.

## How it works

The workflow of `cy.matchImageSnapshot()` when running Cypress is:

1.  Take a screenshot with `cy.screenshot()` named according to the current test.
2.  Check if a saved snapshot exists in `<rootDir>/cypress/snapshots` and if so diff against that snapshot.
3.  If there is a resulting diff, save it to `<rootDir>/cypress/snapshots/__diff_output__`.

## Requirements

Tested on Cypress 10.x, 11.x and 12.x

Cypress must be installed as a peer dependency

## Contributing

### Setup

- Clone the repository and install the yarn dependencies with `yarn install`
- Ensure that Docker is setup. This is necessary for generating/updating snapshots
- Using [Volta](https://volta.sh/) is recommended for managing Node and Yarn versions. These are
  automatically picked up from the `package.json`
- Commits should be based on [conventional-changelog](https://github.com/pvdlg/conventional-changelog-metahub#commit-types)

### Working on the plugin

To make it easier to test whilst developing there are a few simple
Cypress tests that validate the plugin. There are two ways to run these tests:

#### open

`yarn test:open`

In open mode the tests run in Electron and ignore any snapshot failures. This is
due to the rendering differences on developer machines vs CI. Here there is also
verbose output sent to the test runner console to aid debugging.

**Note** here that the yarn script above will re-build the plugin each time. This is
necessary because the tests are run against the output in the `dist` directory
to ensure parity between the built package on NPM.

Ensure that the command is run each time changes need to be tested in Cypress

#### run

- `yarn docker:build`
- `yarn docker:run`

The commands here ensure that the tests are run inside a Docker container that
matches the CI machine. This allows images to be generated and matched correctly
when running the tests in Github Actions.

##### Note on environment variables

It is necessary to have two environment variables defined by default before
running the tests in Docker:

- `CYPRESS_updateSnapshots=false`
- `CYPRESS_debugSnapshots=false`

It's recommended that these are loaded into the shell with something like [direnv](https://direnv.net/)

Then they can be overridden as needed:

```
CYPRESS_updateSnapshots=true yarn docker:run
```

## Forked from `@simonsmith/cypress-image-snapshot`

This is a rewrite of the plugin from [Simon Smith](https://github.com/simonsmith/cypress-image-snapshot).


CHANGELOG.md
# [1.1.0](https://github.com/emerson-eps/cypress-image-snapshot/compare/1.0.2...1.1.0) (2023-06-15)


### Features

* trigger public release ([020e428](https://github.com/emerson-eps/cypress-image-snapshot/commit/020e42819aa33388f9c54879242354f089d0820c))

## [1.0.2](https://github.com/emerson-eps/cypress-image-snapshot/compare/1.0.1...1.0.2) (2023-06-14)


### Bug Fixes

* empty commit ([b982b90](https://github.com/emerson-eps/cypress-image-snapshot/commit/b982b90c8d06b448507602a1618779186a92b964))

## [1.0.1](https://github.com/emerson-eps/cypress-image-snapshot/compare/1.0.0...1.0.1) (2023-06-14)


### Reverts

* Revert "update stuff" ([b2db6a8](https://github.com/emerson-eps/cypress-image-snapshot/commit/b2db6a83306e4d5fcc45f51da9d711f18f62e462))

# [7.0.0](https://github.com/simonsmith/cypress-image-snapshot/compare/6.1.1...7.0.0) (2023-05-25)


### Bug Fixes

* ensure files are packaged in root ([c0816dc](https://github.com/simonsmith/cypress-image-snapshot/commit/c0816dc9b3c809fc31fd9b147a3499a3e4e60f2d))
* move @types/jest-image-snapshot ([5e65567](https://github.com/simonsmith/cypress-image-snapshot/commit/5e65567d2a383f65860976213ebab9a86da3ff72))
* release from root directory ([e0bab6a](https://github.com/simonsmith/cypress-image-snapshot/commit/e0bab6ac3a28d70697cfc2941559b188e6a21cad))


### Features

* add recording of snapshot result ([488ae4b](https://github.com/simonsmith/cypress-image-snapshot/commit/488ae4be65267bb3547064becb864664a24f7846))
* add semantic release ([b1b063b](https://github.com/simonsmith/cypress-image-snapshot/commit/b1b063b3c31b33b25e0fb37e87048533c82a0139))
* allow default options to be passed into addMatchImageSnapshotCommand ([405afcb](https://github.com/simonsmith/cypress-image-snapshot/commit/405afcbd202adcb2665a5239120fb7d0fa02022b))


### BREAKING CHANGES

* removed fork of original package

This is a rewrite of the original library, now with full support for
TypeScript and improved testing.

Notes:

* The API for `matchImageSnapshot` remains the same, as well as all the
  import paths
* The behavior of the plugin is exactly the same, as are the default
  options

TypeScript types are exported under `@simonsmith/cypress-image-snapshot/types`.
These should be used instead of the package on DefinitelyTyped

Removed:
* The `reporter` is not supported in this version.

# [7.0.0-beta.3](https://github.com/simonsmith/cypress-image-snapshot/compare/7.0.0-beta.2...7.0.0-beta.3) (2023-05-24)


### Bug Fixes

* ensure files are packaged in root ([db30cbb](https://github.com/simonsmith/cypress-image-snapshot/commit/db30cbb901b52a88f7959fc1565260fadf3f058e))
* move @types/jest-image-snapshot ([f6404d4](https://github.com/simonsmith/cypress-image-snapshot/commit/f6404d444875efd4e42123dd80e3784c67ec86b1))

# [7.0.0-beta.2](https://github.com/simonsmith/cypress-image-snapshot/compare/7.0.0-beta.1...7.0.0-beta.2) (2023-05-24)


### Bug Fixes

* release from root directory ([0ec36c1](https://github.com/simonsmith/cypress-image-snapshot/commit/0ec36c13bd0ff478ee013f75fc94975a255c33dd))

# [7.0.0-beta.1](https://github.com/simonsmith/cypress-image-snapshot/compare/6.1.1...7.0.0-beta.1) (2023-05-24)


### Features

* add recording of snapshot result ([488ae4b](https://github.com/simonsmith/cypress-image-snapshot/commit/488ae4be65267bb3547064becb864664a24f7846))
* add semantic release ([4db3b89](https://github.com/simonsmith/cypress-image-snapshot/commit/4db3b89690c3e726689ee98f44fa528fcba233e2))
* allow default options to be passed into addMatchImageSnapshotCommand ([405afcb](https://github.com/simonsmith/cypress-image-snapshot/commit/405afcbd202adcb2665a5239120fb7d0fa02022b))


### BREAKING CHANGES

* removed fork of original package

This is a rewrite of the original library, now with full support for
TypeScript and improved tests

Released as a major but everything should be backwards compatible


src/types.ts
import type {MatchImageSnapshotOptions} from 'jest-image-snapshot'

declare global {
  namespace Cypress {
    interface Chainable {
      matchImageSnapshot(
        nameOrCommandOptions?: CypressImageSnapshotOptions | string,
      ): Chainable<DiffSnapshotResult>
      matchImageSnapshot(
        name: string,
        commandOptions: CypressImageSnapshotOptions,
      ): Chainable<DiffSnapshotResult>
    }
  }
}

type CypressScreenshotOptions = Partial<Cypress.ScreenshotOptions>

// The options that are passed around internally from command to plugin
export type SnapshotOptions = {
  screenshotsFolder: string
  isUpdateSnapshots: boolean
  isSnapshotDebug: boolean
  specFileName: string
  currentTestTitle: string
} & CypressScreenshotOptions &
  MatchImageSnapshotOptions &
  ExtraSnapshotOptions

// The options that are exposed to the user via `matchImageSnapshot`
// Prevents the private properties above from showing up in autocomplete
// Merges both Cypress and jest-image-snapshot options together. Not ideal
// if one day they choose a clashing key, but this way it keeps the public
// API non breaking
export type CypressImageSnapshotOptions = Partial<
  CypressScreenshotOptions & MatchImageSnapshotOptions & ExtraSnapshotOptions
>

export type Subject =
  | void
  | Document
  | Window
  | Cypress.JQueryWithSelector<HTMLElement>

export type DiffSnapshotResult = {
  added?: boolean
  receivedSnapshotPath?: string
  updated?: boolean
  imgSrcString: string
  imageDimensions: {
    baselineHeight: number
    baselineWidth: number
    receivedWidth: number
    receivedHeight: number
  }
  pass: boolean
  diffSize: boolean
  diffOutputPath: string
  diffRatio: number
  diffPixelCount: number
}

export type DiffSnapshotOptions = {
  receivedImageBuffer: Buffer
  snapshotIdentifier: string
  snapshotsDir: string
  storeReceivedOnFailure?: boolean
  receivedDir?: string
  diffDir?: string
  updateSnapshot?: boolean
  updatePassedSnapshot?: boolean
  customDiffConfig?: Record<string, unknown>
} & Pick<
  MatchImageSnapshotOptions,
  | 'comparisonMethod'
  | 'blur'
  | 'allowSizeMismatch'
  | 'diffDirection'
  | 'onlyDiff'
  | 'failureThreshold'
  | 'failureThresholdType'
>

type ExtraSnapshotOptions = {
  /**
   * Waits an arbitrary amount of time (in milliseconds) if the baseline image doesn't exist
   * @default 2500
   */
  //TODO: Implement waiting if reference image doesn't exist :)
  //delay: number
  /**
   * Time it takes for the snapshot command to time out if the snapshot is not correct
   * @default 5000
   */
  timeout: number
  /**
   * Sets a delay between recursive snapshots
   * @default 1000.
   */
  delayBetweenTries: number
}


src/command.ts
import path from 'path'
import extend from 'just-extend'
import {MATCH, RECORD} from './constants'
import type {
  CypressImageSnapshotOptions,
  DiffSnapshotResult,
  SnapshotOptions,
  Subject,
} from './types'

const PNG_EXT = '.png'
const SNAP_EXT = `.snap${PNG_EXT}`

const COMMAND_NAME = 'cypress-image-snapshot'
const screenshotsFolder =
  Cypress.config('screenshotsFolder') || 'cypress/screenshots'
const isUpdateSnapshots: boolean = Cypress.env('updateSnapshots') || false
const isSnapshotDebug: boolean = Cypress.env('debugSnapshots') || false

const defaultOptions: SnapshotOptions = {
  screenshotsFolder,
  isUpdateSnapshots,
  isSnapshotDebug,
  specFileName: Cypress.spec.name,
  currentTestTitle: '',
  failureThreshold: 0,
  failureThresholdType: 'pixel',
  timeout: 5000,
  delayBetweenTries: 1000,
}

/**
 * Add this function to your `supportFile` for e2e/component
 * Accepts options that are used for all instances of `toMatchSnapshot`
 */
export const addMatchImageSnapshotCommand = (
  defaultOptionsOverrides: CypressImageSnapshotOptions = {},
) => {
  Cypress.Commands.add(
    'matchImageSnapshot',
    {
      prevSubject: ['optional', 'element', 'document', 'window'],
    },
    matchImageSnapshot(defaultOptionsOverrides),
  )
}

const matchImageSnapshot =
  (defaultOptionsOverrides: CypressImageSnapshotOptions) =>
  (
    subject: Subject,
    nameOrCommandOptions: CypressImageSnapshotOptions | string,
    commandOptions?: CypressImageSnapshotOptions,
  ) => {
    // access the env here so that it can be overridden in tests
    const isFailOnSnapshotDiff: boolean =
      typeof Cypress.env('failOnSnapshotDiff') === 'undefined' || false
    const isRequireSnapshots: boolean = Cypress.env('requireSnapshots') || false

    const {filename, options} = getNameAndOptions(
      nameOrCommandOptions,
      defaultOptionsOverrides,
      commandOptions,
    )

    function recursiveSnapshot():
      | Cypress.Chainable<DiffSnapshotResult>
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      | Cypress.Chainable<DiffSnapshotResult | Cypress.Chainable<any>> {
      const elementToScreenshot = cy.wrap(subject)
      cy.task(MATCH, {
        ...options,
        currentTestTitle: Cypress.currentTest.title,
      })

      const screenshotName = getScreenshotFilename(filename)

      const currentTime = Date.now()
      const hasTimedOut = currentTime - startTime >= Math.abs(options.timeout)

      elementToScreenshot.screenshot(screenshotName, options)
      return cy.task<DiffSnapshotResult>(RECORD).then((snapshotResult) => {
        const {
          added,
          pass,
          updated,
          imageDimensions,
          diffPixelCount,
          diffRatio,
          diffSize,
          diffOutputPath,
        } = snapshotResult

        if (pass) {
          return
        }

        if (added && isRequireSnapshots) {
          const message = `New snapshot: '${screenshotName}' was added, but 'requireSnapshots' was set to true.
            This is likely because this test was run in a CI environment in which snapshots should already be committed.`
          if (isFailOnSnapshotDiff) {
            throw new Error(message)
          } else {
            Cypress.log({name: COMMAND_NAME, message})
            return
          }
        }

        if (!pass && !added && !updated) {
          const message = diffSize
            ? `Image size (${imageDimensions.baselineWidth}x${imageDimensions.baselineHeight}) different than saved snapshot size (${imageDimensions.receivedWidth}x${imageDimensions.receivedHeight}).\nSee diff for details: ${diffOutputPath}`
            : `Image was ${
                diffRatio * 100
              }% different from saved snapshot with ${diffPixelCount} different pixels.\nSee diff for details: ${diffOutputPath}`

          if (isFailOnSnapshotDiff && hasTimedOut) {
            throw new Error(message)
          } else if (hasTimedOut) {
            Cypress.log({name: COMMAND_NAME, message})
          } else {
            Cypress.log({name: COMMAND_NAME, message})
            Cypress.log({
              message: `Attempt ${currentAttempt.toString()} out of ${totalAttempts.toString()}`,
            })
            currentAttempt++
            return cy.wait(options.delayBetweenTries).then(() => {
              return recursiveSnapshot()
            })
          }
        }
      })
    }

    const totalAttempts = Math.floor(
      options.timeout / options.delayBetweenTries,
    )
    let currentAttempt = 1
    const startTime = Date.now()
    recursiveSnapshot()
  }

const getNameAndOptions = (
  nameOrCommandOptions: CypressImageSnapshotOptions | string,
  defaultOptionsOverrides: CypressImageSnapshotOptions,
  commandOptions?: CypressImageSnapshotOptions,
) => {
  let filename: string | undefined
  let options = extend(
    true,
    {},
    defaultOptions,
    defaultOptionsOverrides,
  ) as SnapshotOptions
  if (typeof nameOrCommandOptions === 'string' && commandOptions) {
    filename = nameOrCommandOptions
    options = extend(
      true,
      {},
      defaultOptions,
      defaultOptionsOverrides,
      commandOptions,
    ) as SnapshotOptions
  }
  if (typeof nameOrCommandOptions === 'string') {
    filename = nameOrCommandOptions
  }
  if (typeof nameOrCommandOptions === 'object') {
    options = extend(
      true,
      {},
      defaultOptions,
      defaultOptionsOverrides,
      nameOrCommandOptions,
    ) as SnapshotOptions
  }
  return {
    filename,
    options,
  }
}

const getScreenshotFilename = (filename: string | undefined) => {
  if (filename) {
    return filename
  }
  return Cypress.currentTest.titlePath.join(' -- ')
}

/**
 * replaces forward slashes (/) and backslashes (\) in a given input string with the appropriate path separator based on the operating system
 * @param input string to replace
 */
export const replaceSlashes = (input: string): string => {
  return input.replace(/\\/g, path.sep).replace(/\//g, path.sep)
}
